### PR TITLE
Add hyperLogLogPlus aggregation function for distinct count

### PR DESCRIPTION
**Pull request by deemoliu**
`feature`: add DistinctCountHLLPlus function.

Context:
currently pinot is using Clearspring implementation of HLL algorithm. https://github.com/addthis/stream-lib/blob/master/src/main/java/com/clearspring/analytics/stream/cardinality/HyperLogLog.java
reference: HLL paper: https://algo.inria.fr/flajolet/Publications/FlFuGaMe07.pdf

we have some customers observed ElasticSearch dataset to Pinot, Pinot is using HLL algorithm while ES is using HLL++, user reported that HLL++ has higher accuracy than HLL when cardinality of dimension is at 10k-100k. 
reference: HLL++ paper https://static.googleusercontent.com/media/research.google.com/en//pubs/archive/40671.pdf

We tried to solve the above issue by tuning the log2m parameters of HLL (distinctCountHLL) functions, then we observed the CPU usage increased and bring the cluster into unstable state. 

This PR tried to bridge the gap between ES and Pinot by introducing HyperLogLogPlusPlus to Pinot. Clearspring already has support for the HLL++. https://github.com/addthis/stream-lib/blob/master/src/main/java/com/clearspring/analytics/stream/cardinality/HyperLogLogPlus.java

Approximation Benchmark:

Test Query
```
SELECT
  DISTINCTCOUNTHLLPLUS(some_id, 12)
FROM
  test_pinot_table
where
  time_ms >= 1694380000000
  time_ms <= 1694382000000
  and filter1 > 0
  AND filter2 = 0
  AND filter3 > 0
  AND filter4 != TRUE
```

Approximation Benchmarking result

<meta charset="utf-8"><b style="font-weight:normal;" id="docs-internal-guid-23313897-7fff-1c06-89c2-ee4b9b9994a8"><p dir="ltr" style="line-height:1.38;margin-top:0pt;margin-bottom:0pt;"><span style="font-size:11pt;font-family:Arial,sans-serif;color:#000000;background-color:transparent;font-weight:400;font-style:normal;font-variant:normal;text-decoration:none;vertical-align:baseline;white-space:pre;white-space:pre-wrap;">Result</span></p><div dir="ltr" style="margin-left:0pt;" align="left">

  | Query1 | Query2 | Query3
-- | -- | -- | --
Distinct Count | 144730 | 2797 | 204807
HLL (default log2m=8) | 158486 | 3138 | 191046
HLL log2m=9 | 155441 | 2850 | 209481
HLL log2m=10 | 148871 | 2797 | 204871
HLL++ default p=14 | 143821 | 2787 | 204371
hLL++ p=13 | 145069 | 2774 | 206735
HLL++ p=12 | 145329 | 2770 | 209274
HLL++ p=11 | 145251 | 2720 | 210161
HLL++ p=10 | 147969 | 2679 | 214758
HLL++ p=9 | 147234 | 2767 | 201620
HLL++ p=8 | 141377 | 2653 | 184344

</div></b>

### CODE REVIEW

**Comment by deemoliu** [comment]
@Jackie-Jiang @chenboat @cbalci can you please review this PR?

**Comment by deemoliu** [comment]
Test Query
```
SELECT
  DISTINCTCOUNTHLLPLUS(some_id, 12)
FROM
  test_pinot_table
where
  time_ms >= 1694380000000
  time_ms <= 1694382000000
  and filter1 > 0
  AND filter2 = 0
  AND filter3 > 0
  AND filter4 != TRUE
```

Approximation Benchmarking result

<meta charset="utf-8"><b style="font-weight:normal;" id="docs-internal-guid-23313897-7fff-1c06-89c2-ee4b9b9994a8"><p dir="ltr" style="line-height:1.38;margin-top:0pt;margin-bottom:0pt;"><span style="font-size:11pt;font-family:Arial,sans-serif;color:#000000;background-color:transparent;font-weight:400;font-style:normal;font-variant:normal;text-decoration:none;vertical-align:baseline;white-space:pre;white-space:pre-wrap;">Result</span></p><div dir="ltr" style="margin-left:0pt;" align="left">

  | Query1 | Query2 | Query3
-- | -- | -- | --
Distinct Count | 144730 | 2797 | 204807
HLL (default log2m=8) | 158486 | 3138 | 191046
HLL log2m=9 | 155441 | 2850 | 209481
HLL log2m=10 | 148871 | 2797 | 204871
HLL++ default p=14 | 143821 | 2787 | 204371
hLL++ p=13 | 145069 | 2774 | 206735
HLL++ p=12 | 145329 | 2770 | 209274
HLL++ p=11 | 145251 | 2720 | 210161
HLL++ p=10 | 147969 | 2679 | 214758
HLL++ p=9 | 147234 | 2767 | 201620
HLL++ p=8 | 141377 | 2653 | 184344

</div></b>

File: pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/DistinctCountHLLPlusAggregationFunction.java [hunk 1]
```diff
@@ -0,0 +1,473 @@
...
+public class DistinctCountHLLPlusAggregationFunction extends BaseSingleInputAggregationFunction<HyperLogLogPlus, Long> {
+  // The parameter "p" determines the precision of the sparse list in HyperLogLogPlus.
+  protected final int _p;
+  // The "sp" parameter specifies the number of standard deviations that the sparse list's precision should be set to.
+  protected final int _sp;
+
+  public DistinctCountHLLPlusAggregationFunction(List<ExpressionContext> arguments) {
+    super(arguments.get(0));
+    int numExpressions = arguments.size();
+    // This function expects 1 or 2 arguments.
```
Comment (chenboat) [inline pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/DistinctCountHLLPlusAggregationFunction.java:50]: update javadoc, should we expect 2 or 3 instead based on the code?

Comment (deemoliu) [reply pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/DistinctCountHLLPlusAggregationFunction.java:50]: thank @chenboat for review. this function epxects 1 or 2 or 3 arguments.

**Comment by siddharthteotia** [comment]
I'd like to spend some time reviewing this today / tomorrow.

**Comment by deemoliu** [comment]
> I'd like to spend some time reviewing this today / tomorrow.

@siddharthteotia thanks for reviewing. I will fix this conflict in this hour.

File: pinot-core/src/main/java/org/apache/pinot/core/operator/query/NonScanBasedAggregationOperator.java [hunk 2]
```diff
@@ -233,6 +256,25 @@ private static HyperLogLog getDistinctCountHLLResult(Dictionary dictionary,
...
+      // Treat BYTES value as serialized HyperLogLog
+      try {
+        HyperLogLogPlus hllplus = ObjectSerDeUtils.HYPER_LOG_LOG_PLUS_SER_DE.deserialize(dictionary.getBytesValue(0));
+        int length = dictionary.length();
+        for (int i = 1; i < length; i++) {
+          hllplus.addAll(ObjectSerDeUtils.HYPER_LOG_LOG_PLUS_SER_DE.deserialize(dictionary.getBytesValue(i)));
+        }
+        return hllplus;
+      } catch (Exception e) {
+        throw new RuntimeException("Caught exception while merging HyperLogLogs", e);
```
Comment (jackjlli) [inline pinot-core/src/main/java/org/apache/pinot/core/operator/query/NonScanBasedAggregationOperator.java:271]: nit: `Caught exception while merging HyperLogLogsPlus`

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/aggregator/DistinctCountHLLPlusValueAggregator.java [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+  @Override
+  public HyperLogLogPlus cloneAggregatedValue(HyperLogLogPlus value) {
+    return deserializeAggregatedValue(serializeAggregatedValue(value));
+  }
+
+  @Override
+  public int getMaxAggregatedValueByteSize() {
+    // NOTE: For aggregated metrics, initial aggregated value might have not been generated. Returns the byte size
+    //       based on log2m.
+    return _maxByteSize > 0 ? _maxByteSize : 100000000;
```
Comment (jackjlli) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/aggregator/DistinctCountHLLPlusValueAggregator.java:113]: This line doesn't seem to align with the comment above?

**Review by jackjlli — COMMENTED** [review]
Thanks for working on this! Would you mind adding more info to this PR, like what functionalities are added, what's the difference between the raw one and the non-raw one, what the purpose of your benchmarking test is?

File: pinot-core/src/main/java/org/apache/pinot/core/common/ObjectSerDeUtils.java [hunk 4]
```diff
@@ -554,6 +559,35 @@ private HyperLogLog deserialize(IntBuffer intBuffer) {
...
+      try {
+        return HyperLogLogPlus.Builder.build(bytes);
+      } catch (IOException e) {
+        throw new RuntimeException("Caught exception while serializing HyperLogLogPlus", e);
+      }
+    }
+
+    @Override
+    public HyperLogLogPlus deserialize(ByteBuffer byteBuffer) {
+      // NOTE: The passed in byte buffer is always BIG ENDIAN
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/common/ObjectSerDeUtils.java:584]: (minor) Remove this comment as it doesn't apply

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/utils/CustomSerDeUtils.java [hunk 5]
```diff
@@ -188,6 +189,38 @@ public HyperLogLog deserialize(ByteBuffer byteBuffer) {
...
+
+    @Override
+    public HyperLogLogPlus deserialize(ByteBuffer byteBuffer) {
+      byte[] bytes = new byte[byteBuffer.remaining()];
+      byteBuffer.get(bytes);
+      try {
+        return HyperLogLogPlus.Builder.build(bytes);
+      } catch (IOException e) {
+        throw new RuntimeException("Caught exception while de-serializing HyperLogLogPlus", e);
+      }
```
Comment (Jackie-Jiang) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/utils/CustomSerDeUtils.java:220]:
```suggestion
      return deserialize(bytes);
```

**Review by Jackie-Jiang — APPROVED** [review]
LGTM with minor comments

**Comment by deemoliu** [comment]
thanks @Jackie-Jiang @jackjlli @siddharthteotia and @chenboat for review. I will address comment today.

File: pinot-core/src/main/java/org/apache/pinot/core/operator/query/NonScanBasedAggregationOperator.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (deemoliu) [reply pinot-core/src/main/java/org/apache/pinot/core/operator/query/NonScanBasedAggregationOperator.java]: thanks @Jackie-Jiang i fixed the error message.

**Review by jackjlli — APPROVED** [review]
Minor but LGTM. Thanks for adding the aggregation function!

File: pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/DistinctCountHLLPlusAggregationFunction.java [hunk 6]
```diff
@@ -0,0 +1,473 @@
...
+    BlockValSet blockValSet = blockValSetMap.get(_expression);
+
+    // Treat BYTES value as serialized HyperLogLogPlus
+    DataType storedType = blockValSet.getValueType().getStoredType();
+    if (storedType == DataType.BYTES) {
+      byte[][] bytesValues = blockValSet.getBytesValuesSV();
+      try {
+        HyperLogLogPlus hyperLogLogPlus = aggregationResultHolder.getResult();
+        if (hyperLogLogPlus != null) {
+          for (int i = 0; i < length; i++) {
```
[the remark this answers is not in the document]
Comment (deemoliu) [reply pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/DistinctCountHLLPlusAggregationFunction.java]: thanks for code review! addressed the code comment, please take a look

File: pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/DistinctCountHLLPlusAggregationFunction.java [hunk 7]
```diff
@@ -96,16 +96,12 @@ public void aggregate(int length, AggregationResultHolder aggregationResultHolde
...
-          }
-        } else {
+        if (hyperLogLogPlus == null) {
           hyperLogLogPlus = ObjectSerDeUtils.HYPER_LOG_LOG_PLUS_SER_DE.deserialize(bytesValues[0]);
           aggregationResultHolder.setValue(hyperLogLogPlus);
-          for (int i = 1; i < length; i++) {
-            hyperLogLogPlus.addAll(ObjectSerDeUtils.HYPER_LOG_LOG_PLUS_SER_DE.deserialize(bytesValues[i]));
-          }
+        }
+        for (int i = 0; i < length; i++) {
```
Comment (jackjlli) [inline pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/DistinctCountHLLPlusAggregationFunction.java:103]: nit: `int i = 1;`

File: pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/DistinctCountHLLPlusAggregationFunction.java [hunk 8]
```diff
@@ -0,0 +1,469 @@
...
+    // Treat BYTES value as serialized HyperLogLogPlus
+    DataType storedType = blockValSet.getValueType().getStoredType();
+    if (storedType == DataType.BYTES) {
+      byte[][] bytesValues = blockValSet.getBytesValuesSV();
+      try {
+        HyperLogLogPlus hyperLogLogPlus = aggregationResultHolder.getResult();
+        if (hyperLogLogPlus == null) {
+          hyperLogLogPlus = ObjectSerDeUtils.HYPER_LOG_LOG_PLUS_SER_DE.deserialize(bytesValues[0]);
+          aggregationResultHolder.setValue(hyperLogLogPlus);
+        }
```
Comment (jackjlli) [inline pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/DistinctCountHLLPlusAggregationFunction.java:102]: We need the else block if the check is not qualified:
```
} else {
    hyperLogLogPlus.addAll(ObjectSerDeUtils.HYPER_LOG_LOG_PLUS_SER_DE.deserialize(bytesValues[0]));
}
```

Comment (deemoliu) [reply pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/DistinctCountHLLPlusAggregationFunction.java:104]: got your point, updated, thanks!

**Pull request merged**